### ISSUE TITLE
ICE when reading noinline functions in unsafe context

### ISSUE THREAD

**Issue by Munksgaard** [diku-dk/futhark#1366]
```futhark
#[noinline]
let update [n] (xs: [n]i32): [n]i32 =
  map (+ 1) xs

let main [n][m] (xss: *[m][n]i32) =
  #[unsafe]
  loop xss for i < m do
    copy xss with [i] = update xss[i]
```

To reproduce:

```
$ futhark dev --kernels -a -e --cse -e ~/tmp/ice.fut  > ~/tmp/ice.fut_kernels_mem

$ futhark dev -e ~/tmp/ice.fut_kernels_mem
Internal compiler error.
Please report this at https://github.com/diku-dk/futhark/issues.
Type error after pass 'simplify':
In function entry_main
When checking function body
In expression of statement
  {main_res_mem_4316 : ({}, mem);
   main_res_4265 : ({}, [m_4263][n_4262]i32 @ main_res_mem_4316 ->
                    {base: [m_4263, n_4262]; contiguous: true;
                     LMADs: [{offset: 0i64; strides: [n_4262, 1i64];
                              rotates: [0i64, 0i64]; shape: [m_4263, n_4262];
                              permutation: [0, 1]; monotonicity: [Inc, Inc]}]})}
Inside the loop body
In expression of statement
  {lw_val_mem_4302 : ({}, mem);
   lw_val_4269 : ({n_4262, mem_4300, update_arg_linear_4301},
                  [n_4262]i32 @ lw_val_mem_4302 ->
                  {base: [n_4262]; contiguous: true; LMADs: [{offset: 0i64;
                                                              strides: [1i64];
                                                              rotates: [0i64];
                                                              shape: [n_4262];
                                                              permutation: [0];
                                                              monotonicity: [Inc]}]})}
Call of unknown function update_4129<unsafe>.
```

**Comment by athas** [comment]
This is embarrassing for many reasons, but one of those is that we don't run the type checker after reading a core program.  I suspect that makes the test suite less sensitive than it should be.

**Comment by athas** [comment]
Closed in 053e4e34461b6f73671e2ff02c8c2f9f35ed8e6e.
